### ISSUE TITLE
Typing latency

### ISSUE THREAD

**Issue by indutny** [zeit/hyper#296]
Is there any way to reduce latency when typing things? It is not super slow, but noticeably slower than any other terminal that I use.

It feels like I'm typing on ssh terminal.

**Comment by matheuss** [comment]
I don't see that latency here 😰

**Comment by rauchg** [comment]
Neither do I. @indutny can you record a video with your mobile phone?

**Comment by iamstarkov** [comment]
there are tools to objective latency measurement
https://pavelfatin.com/typing-with-pleasure/
in russian http://tonsky.livejournal.com/306545.html

**Comment by indutny** [comment]
@iamstarkov 👍

**Comment by indutny** [comment]
Just wrote a small C application to measure latency. Turns out HyperTerm is actually faster than iTerm2. My bad!

Still would be cool if anyone else would be able to confirm it independently. I am not absolutely sure that my tool is correct.

(The tool is [here][0])

<img width="602" alt="screen shot 2016-07-26 at 10 01 39 pm" src="https://cloud.githubusercontent.com/assets/238531/17161567/90e31942-537c-11e6-9f38-8ef07834d481.png">

[0]: https://gist.github.com/indutny/b530593a996bbbe8bfd51fc3f38c45f2

**Comment by indutny** [comment]
Here is updated bench with the latency of typing in this comments form (just for fun):
<img width="602" alt="screen shot 2016-07-26 at 10 04 21 pm" src="https://cloud.githubusercontent.com/assets/238531/17161622/f98a7fc6-537c-11e6-9c45-d8546b2f6829.png">

**Comment by indutny** [comment]
Though, may be HyperTerm just became faster since last time? :) Have you done anything about performance?

**Comment by indutny** [comment]
@matheuss I already did in a comment above. Here it is just in case: https://gist.github.com/indutny/b530593a996bbbe8bfd51fc3f38c45f2

**Comment by matheuss** [comment]
Oh, my bad 😅 Thanks!

**Comment by freebroccolo** [comment]
> Still would be cool if anyone else would be able to confirm it independently.

My observations (subjective) are that HyperTerm has noticeably less latency as I type than iTerm2 but in my case I know a lot of that is due to my zsh config which causes a lot of redrawing as I type.

> Though, may be HyperTerm just became faster since last time? :) Have you done anything about performance?

There have been a couple of updates of `electron` in the HyperTerm package. The most recent one `1.3.0` updates Chrome to version `52` and node to `6.3.0`. So there may be some performance improvements just from that.

**Comment by rauchg** [comment]
There are certainly *some* actions at which HyperTerm can be slow. Typing in a new prompt is certainly not one of them (as shown above :P), but we have to consider a lot of different scenarios (like large screen redraws).

**Comment by indutny** [comment]
@freebroccolo 

> There have been a couple of updates of electron in the HyperTerm package. The most recent one 1.3.0 updates Chrome to version 52 and node to 6.3.0. So there may be some performance improvements just from that.

This must be it. I wish I had an old version to compare to.

**Comment by indutny** [comment]
Decided to publish my tool as a separate repo with build instructions (OS X only): https://github.com/indutny/latetyper

**Comment by matheuss** [comment]
You can find the old versions [here](https://github.com/zeit/hyperterm/releases) ^~^

**Comment by TooTallNate** [comment]
Incredible @indutny

**Comment by indutny** [comment]
@matheuss just tested old versions, all of them seems to be faster than the latest iTerm2 :(

**Comment by indutny** [comment]
Good job HyperTerm, though!

**Comment by samccone** [comment]
> Is there any way to reduce latency when typing things?

So it comes down to our measurement approach here. The approach. I chose here is "when can I see the letters after hitting a button on my keyboard". If we drop down into devtools and look at things per each frame we can get a better look at exactly when the new chars are shipped to the user.

![image](https://cloud.githubusercontent.com/assets/883126/17198008/94a469dc-5426-11e6-946c-fbd46b90f560.png)

^ We can see the majority of our frames are shipped to the user well under budget. Since the update loop for the screen is running inside of a requestAnimationFrame we are kinda self limiting how fast we are updating the screen (this is good, and bad in someways -- if we are trying to optimize for shipping the chars to screen as fast as possible)

![image](https://cloud.githubusercontent.com/assets/883126/17197978/4acf0812-5426-11e6-9129-ced1d8c86a9b.png)

I will be looking more into the costs associated with typing and some easy wins that we can get. 

More on this later ✌️

**Comment by albinekb** [comment]
Nice @samccone,
This might be relevant: http://www.html5rocks.com/en/tutorials/speed/unnecessary-paints/

**Comment by mike-engel** [comment]
I've noticed some "perceived" latency when using the backspace key—that is, when typing, it seems pretty performant. When using backspace/delete, I've noticed that there's a small time difference between when the character goes away and when the cursor moves to the previous position. I don't know if this is what @indutny was talking about (it may have been typoing in both directions). If so, we can probably do some work to move the cursor faster to where it should be.

**Comment by samccone** [comment]
> goes away and when the cursor moves to the previous position

This is due to how hterm implements the cursor update :)
https://github.com/dbkaplun/hterm-umdjs/blob/master/dist/index.js#L13004-L13013

I would say this is WAI :)

**Comment by mike-engel** [comment]
Interesting...thanks @samccone. Probably some room for improvement there :smile:

**Comment by samccone** [comment]
@chenglou do you observe "latency" in the char being deleted or the cursor moving?

**Comment by chenglou** [comment]
The cursor moving. Also somewhat visible during insertion.

Here's a slowed down recording of what I'm seeing:
![perceivedlatency](https://cloud.githubusercontent.com/assets/1909539/17652437/c51131e6-6230-11e6-9820-e832feb9f6b5.gif)

**Comment by samccone** [comment]
> This is due to how hterm implements the cursor update :)
> https://github.com/dbkaplun/hterm-umdjs/blob/master/dist/index.js#L13004-L13013

^ 😄  -- this could be fixed upstream

**Comment by Zirro** [comment]
I can confirm seeing the delay between character input and cursor movement as well. Unfortunately it is enough to make typing in HyperTerm less pleasant for me.

**Comment by reicheltp** [comment]
I'm not sure if this belongs here or if I should open another issue, but:

If I am using render heavy commandline apps, e.g. midnight commander. I can recognize a very heavy latency of half a second with every input. This make it unusable. 

You can reproduce this with installing midnight commander and just try to navigate with the arrow keys.

**Comment by samccone** [comment]
@reicheltp can you record a timeline please right click and save and then share a link to it on https://chromedevtools.github.io/timeline-viewer/

thanks

**Comment by rauchg** [comment]
@reicheltp very interesting! I'd love to reproduce it as well. Can you tell me exactly what programs you open and in what order?

**Comment by reicheltp** [comment]
@samccone Here a short timeline:
https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://gist.github.com/reicheltp/0bd72cbe69575b43073c7b02320b4bc0/raw

The first long frame is, when I opened the midnight commander with `mc`, Second and third are a simple `tab` to switch columns. And fifth and sixth are `page-up` and `page-down`. As you can see, the frames are really long. 

@rauchg I have used Midnight Commander for Windows (Get it here https://sourceforge.net/projects/mcwin32/) or the OSX/Linux version here (https://www.midnight-commander.org/)

My Setup is Windows 10 Enterprise x64 with latest Hyper.app. Tried with powershell and cmd.

**Comment by chabou** [comment]
@reicheltp is it still a valid issue? Can you reproduce with v1.3.1?

**Comment by reicheltp** [comment]
@chabou Although I can notice a slightly performance improvement with v1.3.1, the issue still exists.

**Comment by simoncpu** [comment]
I've just installed Hyper a while ago, and while I don't have any concrete data to back this up, I can definitely _feel_ the typing delay.

**Comment by albinekb** [comment]
Without plugin? What OS/Hardware? @simoncpu I can't feel it on my Macbook Pro 2017 running High Sierra, but that could be that my computer is overpowered 👍

**Comment by simoncpu** [comment]
@albinekb Yes, without plugin. OS is Windows 10 Pro, hardware is Thinkpad P51. I currently settled for wsltty because Bash on Windows works out of the box (I previously used ConEmu, but had problems with the colors).

BTW, I've just tried Hyper right now (normal command prompt, not bash), and there's a few milliseconds of delay between pressing the keyboard and displaying the characters on the screen. It's not a deal breaker, and I think most people won't notice it, but the experience is akin to typing commands in an SSH session.

**Comment by albinekb** [comment]
Thanks for that @simoncpu, we're currently working on replacing the terminal rendering lib to xterm.js, which will give better results. After that there's lots more we can optimize to make it faster! 🎉

**Comment by simoncpu** [comment]
I think I can feel (sorry, no hard data, just gut feel hehe) less lag when I right clicked on the Hyper icon and chose:

Run with graphic processor > Integrated graphics

The default is "High performance NVIDIA processor". Electron issue with NVIDIA, maybe?

**Comment by vmourac** [comment]
Hi,
In my system the issue happens when I'm using `hyperline#0.6.1`...
As soon as I disable it in `.hyper.js` and restart the app, the latency vanishes

**Comment by balcsida** [comment]
Hi all,

I had experienced lags and high CPU usage on Windows 10 as well and it was resolved by adding `%APPDATA%/hyper` folder as an exclusion to all of my installed antivirus (Windows Defender and Malwarebytes in my case).
I believe this happens because Hyper is built on Electron and the current AV solutions thinks as a "browser".

**Comment by AustinTruex68** [comment]
I have tried what was posted above.. this never happened until I updated to 2.0. Does anyone else have some ideas as to what causes this latency? I don't wanna use cmd.

**Comment by chabou** [comment]
@AustinTruex68 are you on Windows?
This issue is quite outdated.

Maybe you have same problem than https://github.com/zeit/hyper/issues/2838?
If not please open a new issue with all needed information to reproduce

**Comment by AustinTruex68** [comment]
@chabou, sorry about that. Yes it is the same issue as #2838

**Comment by offero** [comment]
@freebroccolo Can you elaborate more on the zsh issue? I am experiencing latency as well and want to figure it out.

**Comment by pgould94** [comment]
I'm on windows and I was also having a lot of latency issues with it. I was using with Gitbash (MINGW64) and if I opened up a new window and used cd, mkdir, ls etc. it was totally fine. However when I opened up node apps using inquirer, cl-table, etc. It would be so bad that scrolling would be impossible and there would be a 1 second lag typing in the prompts.

I'm really sad because I had a theme that matched my vscode which I really liked. Oh well.
